lib/spack/spack/repository.py

```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
         self._all_package_names = None
         self._provider_index = None
 
-        # If repo_dirs is empty, just use the configuration
-        if not repo_dirs:
-            import spack.config
-            repo_dirs = spack.config.get_config('repos')
-            if not repo_dirs:
-                raise NoRepoConfiguredError(
-                    "Spack configuration contains no package repositories.")
-
         # Add each repo to this path.
         for root in repo_dirs:
-            try:
-                repo = Repo(root, self.super_namespace)
-                self.put_last(repo)
-            except RepoError as e:
-                tty.warn("Failed to initialize repository at '%s'." % root,
-                         e.message,
-                         "To remove the bad repository, run this command:",
-                         "    spack repo rm %s" % root)
+            self.append_from_path(root)
+
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```

Comment:
I should preface by saying I'm not sure why ```RepoPath.swap``` ever had to exist - why couldn't ```conftest.builtin_mock``` have said something like

```
saved_repo = spack.repo
spack.repo = mock_db
```

If something more complex is necessary, and if the MutableSequence functionality makes this easier, then I would like the change more.

In my opinion inheriting from MutableSequence makes it harder to modify because it's harder to understand how it works. It is true that an IDE may retrieve the appropriate function, but you still have to mentally combine this sequence of function calls into the desired end result.

> It does make the class easier to use, as it is now adhering to a stronger contract

Speaking generically, enforcing the stronger contract is part of what makes it harder to modify: when/if someone edits RepoPath they must consider whether their operations satisfy the requirements placed on a MutableSequence.

It can be worthwhile to increase complexity if other things are made simpler, but the only simplification I saw was eliminating the call to ```.swap```. In my opinion there ought to be a way to achieve that simplification that would avoid introducing the complexity of satisfying an additional contract.